pysipp/launch.py

```diff
@@ -157,32 +156,32 @@ def clear(self):
         self._procs.clear()
 
 
-async def run_all_agents(runner, agents, timeout=180):
-    """Run a sequencec of agents using a ``TrioRunner``.
-    """
-    async def finalize():
-        # this might raise TimeoutError
-        cmds2procs = await runner.get(timeout=timeout)
-        agents2procs = list(zip(agents, cmds2procs.values()))
-        msg = report.err_summary(agents2procs)
-        if msg:
-            # report logs and stderr
-            await report.emit_logfiles(agents2procs)
-            raise SIPpFailure(msg)
-
-        return cmds2procs
+async def run_all_agents(runner, agents, timeout=180, block=True):
+    """Run a sequencec of agents using a ``TrioRunner``."""
 
     try:
-        await runner.run(
-            (ua.render() for ua in agents),
-            timeout=timeout
-        )
-        await finalize()
+        await runner.run((ua.render() for ua in agents), timeout=timeout)
+        if block:
+            await finalize(runner, agents, timeout)
         return runner
     except TimeoutError as terr:
         # print error logs even when we timeout
         try:
-            await finalize()
+            await finalize(runner, agents, timeout)
         except SIPpFailure as err:
-            assert 'exit code -9' in str(err)
+            assert "exit code -9" in str(err)
             raise terr
+
+
+async def finalize(runner, agents, timeout):
+    """Block up to `timeout` seconds for all agents to complete."""
+    # this might raise TimeoutError
+    cmds2procs = await runner.get(timeout=timeout)
```

Comment:
Hmm could be maybe do this in more *trionic* style using `with trio.fail_after(timeout):` ?

Not sure what will have to get factored out of `.get()`

Reply:
i will give it a try and see if we can refactor `finalize()` and `.get()`

Reply:
Yah I mentioned the spot from my original branch [here](https://github.com/SIPp/pysipp/pull/55#discussion_r554251994). Not sure if I'm convinced though given a lot of users might not be familiar with this style?

What do you think?

Reply:
oh yes i would prefer a common global timeout, we can also document it

> users might not be familiar with this style

which style are you referring to?

Reply:
```python
with trio.fail_after(1):
    scen.finalize()
```

I'm actually now thinking we can't do this anyway..
If we want to keep a sync api this won't work I don't think (`fail_after()` won't work outside a call to `trio.run()` pretty sure). I just don't have my head in this code so I'm probably making bad suggestions.

This *should work* if we offer an async `.aclose()` method to do finalization.

Reply:
Oh wait, nm `.get()` [is async now](https://github.com/SIPp/pysipp/pull/55/files/6394960068455f9d040a31691c568015b2911910#diff-2b80e3f62bb560133d08a5a07a74bc8eac811db3cdf2680136050b78f37d324dR72) - see what I mean, obviously haven't looked at this in a while.

And, duh, we `await` that in the above code. 
So yeah I think maybe pulling out the timeout it the more trionic way to do this thus giving the user more control on waiting / cancellation. The only down side is we'll have to document this style for async users.